File: obj2mesh.js
"use strict";

//very simple obj2mesh: No materials; no vertex condensing,
//no triangulation

var fs = require("fs");

function Writer(fname){
    this.stream = fs.createWriteStream(fname);
    this.offset=0;
}

Writer.prototype.write = function(x){
    //might not work if x is a string with international characters
    this.stream.write(x);
    this.offset += x.length;
}

Writer.prototype.tell = function(){
    return this.offset;
}

Writer.prototype.end = function(x){
    this.stream.end(x);
}

function sub(a,b,len){
	var c = [];
	len = len !== undefined ? len : 3;
	for(var i = 0; i < len; ++i)
		c.push(a[i]-b[i]);
	return c;
}

function main(){
    var infile = process.argv[2];
    var outfile = infile.split(".");
    outfile.pop();
    outfile.push(".mesh");
    for(var i = 1; i < outfile.length; ++i)
    	outfile[0] += outfile[i];
    outfile = outfile[0];
    
    var objdata = fs.readFileSync(infile,{encoding:"utf8"});
    objdata = objdata.split("\n");
    

    var vertexdata = [];
    var triangles = [];
    var texturedata = [];
    var vdata = [], idata = [];
    var vmap = {}, mdict = {};
    var nv = 0;
    var currmtl;
    var objects = [];
    var numVerts = [];
    var xSum = [], ySum = [], zSum = [];
    var normalData = [];
    
    console.log("Reading " + infile + "...");
    for(var oi=0;oi<objdata.length;++oi){
        var L = objdata[oi].split(" ");
        if( L[0] === "o" ){
        	objects.push(L[1]);
        	numVerts.push(0);
        	xSum.push(0);
        	ySum.push(0);
        	zSum.push(0);
        }
        else if( L[0] === "v" ){
        	vertexdata.push([parseFloat(L[1]), parseFloat(L[2]), parseFloat(L[3])]);
        	++numVerts[objects.length-1];
        	xSum[objects.length-1] += vertexdata[vertexdata.length-1][0];
        	ySum[objects.length-1] += vertexdata[vertexdata.length-1][1];
        	zSum[objects.length-1] += vertexdata[vertexdata.length-1][2];
        }
        else if(L[0] === "vn"){
        	normalData.push([parseFloat(L[1]), parseFloat(L[2]), parseFloat(L[3])]);
        }
        else if(L[0] === "vt"){
        	texturedata.push([parseFloat(L[1]), parseFloat(L[2])]);
        }
        else if( L[0] === "f" ){
        	var t = [];
            if( L.length !== 4 )
                console.log("Warning: Non-triangles present");
            
            for(var i=1;i<4;++i){
                var tmp = L[i].split("/");
                var ti = tmp[1];
                var vi=parseInt(tmp[0],10)-1;
                var ni = tmp[2];
                
                if(ti === undefined || ti.length === 0){
                	throw new Error("No texture coordinates");
                }else{
                	ti = parseInt(ti,10)-1;
                }
                if(ni === undefined || ni.length === 0)
                	throw new Error("No normal data");
                else
                	ni = parseInt(ni,10)-1;
                
                t.push(vi, ti, ni);
            }
            triangles.push(t);
        }
        else if(L[0] === "mtllib"){
        	var ml = fs.readFileSync(L[1], {encoding:"utf8"});
        	var mname;
        	ml = ml.split("\n");
        	for(var m = 0; m < ml.length; m++){
        		var tmp = ml[m].split(" ");
        		if(tmp[0] === "newmtl"){
        			mname = tmp[1];
        			mdict[mname] = {};
        		}
        		else if(tmp[0] === "map_Kd")
        			mdict[mname].map_Kd = tmp[1];
        		else if(tmp[0] === "map_Bump")
        			mdict[mname].map_Bump = tmp[1];
        	}
        }
        else if(L[0] === "usemtl"){
        	currmtl = L[1];
        }
    }
    
    console.log("Recording verts and indices...")
    for(var i = 0; i < triangles.length; i++){
    	var t = triangles[i];
    	for(var j = 0; j < 3; ++j){
    		var vi = t[j*3];
    		var ti = t[j*3+1];
    		var ni = t[j*3+2];
    		var key = vi+","+ti+","+ni;
    		if(vmap[key] === undefined){
    			vmap[key]=nv;
    			vdata.push(
    					vertexdata[vi][0],
    					vertexdata[vi][1],
    					vertexdata[vi][2],
    					texturedata[ti][0],
    					texturedata[ti][1],
    					normalData[ni][0],
    					normalData[ni][1],
    					normalData[ni][2],
    					0,0,0 //tangent placeholder
    					);
    			nv++;
    		}
    		idata.push(vmap[key]);
    	}
    }
    
    console.log("Computing tangent vectors...");
    for(var triIdx = 0; triIdx < idata.length; ++triIdx){
    	//q,r,s = triangle vertices
    	var qrsIdx = [idata[triIdx] * 11, idata[++triIdx] * 11, idata[++triIdx] * 11];
    	var q = vdata.slice(qrsIdx[0], qrsIdx[0]+5);
    	var r = vdata.slice(qrsIdx[1], qrsIdx[1]+5);
    	var s = vdata.slice(qrsIdx[2], qrsIdx[2]+5);
    	var qTex = [0, q.pop()];
    	var rTex = [0, r.pop()];
    	var sTex = [0, s.pop()];
    	
    	qTex[0] = q.pop();
    	rTex[0] = r.pop();
    	sTex[0] = s.pop();
    	
    	r = sub(r, q); //R'xyz
    	s = sub(s, q); //S'xyz
    	rTex = sub(rTex, qTex, 2); //R'st
    	sTex = sub(sTex, qTex, 2); //S'st
    	
    	/*
    	 * [Tx Ty Tz]     [R's R't]^-1    [R'x R'y R'z]
    	 * [Bx By Bz]  =  [S's S't]    *  [S'x S'y S'z]
    	 * 
    	 * A = 1/(R's*S't-R't*S's) [S't -R't]
    	 * 
    	 * [Tx Ty Tz]  =  A  *  [R'x R'y R'z]
    	 *                      [S'x S'y S'z]
    	 */
    	
    	var tmp = 1/(rTex[0] * sTex[1] - rTex[1] * sTex[0]);
    	var A = [tmp * sTex[1], tmp * -rTex[1]];
    	var T = [
    	         A[0] * r[0] + A[1] * s[0],
    	         A[0] * r[1] + A[1] * s[1],
    	         A[0] * r[2] + A[1] * s[2]
    	         ];
    	
    	for(var i = 0; i < 3; ++i){
    		vdata[qrsIdx[0]+8+i] += T[i];
    		vdata[qrsIdx[1]+8+i] += T[i];
    		vdata[qrsIdx[2]+8+i] += T[i];
    	}
    }
    
    for(var i = 0; i < vdata.length; i+=11){
    	var xyz = vdata.slice(i+8, i+11);
    	var len = Math.sqrt(xyz[0]*xyz[0] + xyz[1]*xyz[1] + xyz[2]*xyz[2]);
    	for(var j = 0; j < 3; ++j)
    		vdata[i+8+j] = xyz[j] / len;
    }
    
    //output geometry data
    console.log("Writing to file...");
    var ofp = new Writer(outfile);

    ofp.write("mesh_4\n");
    
    if(mdict[currmtl].map_Kd !== undefined)
    	ofp.write("texture_file " + mdict[currmtl].map_Kd + "\n");
    if(mdict[currmtl].map_Bump !== undefined)
    	ofp.write("normal_map " + mdict[currmtl].map_Bump + "\n")
    
    for(var i = 0; i < objects.length; ++i){
    	ofp.write("object " + objects[i] + " " + numVerts[i] + " " + xSum[i]/numVerts[i] + "," + ySum[i]/numVerts[i] + "," + zSum[i]/numVerts[i] + "\n");
    }
    
    //first, write all vertices
    
    //this is count of num floats per vertex * num vertices 
    ofp.write("vertices "+vdata.length+"\n");
    ofp.write("indices "+idata.length+"\n");

    ofp.write("vertex_data");
    //must pad to 4 byte boundary
    while( (ofp.tell()+1) % 4 !== 0 )
        ofp.write(" ");
    ofp.write("\n");
    
    //write all vertices
    var b = new Buffer(vdata.length*4);
    for(var i=0;i<vdata.length;++i){
        b.writeFloatLE(vdata[i],i*4);
    }
    ofp.write(b);
       
    ofp.write("index_data");
    while( (ofp.tell()+1) % 4 !== 0 )
        ofp.write(" ");
    ofp.write("\n");

    //next, write all triangles
    var b = new Buffer(idata.length*2);
    for(var i=0;i<idata.length;++i){
        b.writeUInt16LE(idata[i],i*2);
    }
    ofp.write(b);
    ofp.write("end\n");
    ofp.end();
    console.log(outfile + " successfully created!");
}

main()

File: converter/obj2mesh.py
#!/usr/bin/env python3

import sys
import array
import os
import math 

def sub(a, b, length = 3):
	c = []
	for i in range(0, length):
		c.append(a[i]-b[i])
	return c

def main():
    
    #get input and output file names
    instem = sys.argv[1]
    infile = instem+".obj"
    inrig = instem+".rig"
    outfile = infile+".mesh"
    
    #read obj file as a list of strings
    objdata = open(infile).readlines()
    
    #vertex data will be a list of [x,y,z] tuples
    vertexdata = [] 
    
    #list of [s,t] tuples
    texturedata = [] 
    
    #list of [x,y,z] tuples
    normaldata = []
    
    #list of triangles. Each triangle consists of three 
    #(vertex index, texture index) tuples
    triangles = [] 
    
    #materials. Key = material name; value = dictionary
    mdict = {} 
    
    #current material
    currmtl=None
    
    #information about each object (centroids, etc)
    objects={}
    
    for L in objdata:
        #get next line from input
        L=L.strip().split(" ")
        
        if L[0] == "o" :
            currobj = L[1]
            objects[currobj]={}
            objects[currobj]["centroid"]=[0,0,0]
            objects[currobj]["numv"]=0
        elif L[0] == "v" :
            #a vertex (xyz) specification
            #format: x,y,z w,w,w,w i,i,i,i
            #where the w's are bone weights and the
            #i's are bone indices.
            #If these are absent, dummy values are used.
            tmp = [float(qq) for qq in L[1:]]
            if len(tmp) == 3:
                tmp += [1.0,0.0,0.0,0.0, 1.0,0.0,0.0,0.0]
            elif len(tmp) == 11:
                pass
            else:
                print("Bad vertex:",L)
            vertexdata.append(tmp) 
            for iii in range(3):
                objects[currobj]["centroid"][iii] += vertexdata[-1][iii]
            objects[currobj]["numv"] += 1
        elif L[0] == "vt" :
            #texture coordinate
            texturedata.append([  float(L[1]),
                                float(L[2])]) 
        elif L[0] == "vn":
            #normal
            normaldata.append( [float(L[1]),float(L[2]),float(L[3])] )
        elif L[0] == "f" :
            #face (triangle)
            
            if len(L) != 4:
                print("Warning: Non-triangles present:",len(L)-1,"vertices") 
                continue 
            
            #t=the current triangle
            t=[] 
            for i in range(1,4):
                #four formats possible: vi  vi//ni  vi/ti  vi/ti/ni
                tmp = L[i].split("/") 
                
                #obj uses one-based indexing, so we must subtract one here
                vi=int(tmp[0])-1 
                
                #if no texture coordinate, make up one
                if len(tmp) < 2 or len(tmp[1]) == 0:
                    ti=0
                else:
                    ti = int(tmp[1])-1 
                    
                if len(tmp) < 3 or len(tmp[2]) == 0:
                    ni=0
                else:
                    ni = int(tmp[2])-1
                    
                t.append( (vi,ti,ni) )
                
            triangles.append(t) 
            
        elif L[0] == "mtllib" :
            #material library; must parse it
            ML = open(L[1]).readlines()
            
            #look at each material and store information about it
            for m in ML:
                tmp=m.strip().split(" ")
                if tmp[0] == "newmtl" :
                    mname = tmp[1] 
                    mdict[mname]={}
                elif tmp[0] == "map_Kd" :
                    mdict[mname]["map_Kd"] = tmp[1] 
                elif tmp[0] == "map_Bump" or tmp[0] == "bump" :
                    mdict[mname]["map_Kd"] = tmp[1] 
                elif tmp[0] == "map_Ks" :
                    mdict[mname]["map_Ks"] = tmp[1] 
                
        elif L[0] == "usemtl" :
            #change currently active material
            currmtl = L[1] 
    
    
    #if object lacks texture coordinates, make sure we don't
    #get an out-of-bounds error from accessing texture data
    if len(texturedata) == 0 :
        texturedata += [0,0]
 
    #first, determine how many unique vertices we'll have
    vmap={}     #key=vi,ti,ni  Value=index in vdata
    vdata=[]    #x,y,z,s,t for each vertex, one after another
    idata=[]    #triangle indices: Refers to entries in vdata
    vsize=0
    
    for T in triangles:
        #T will be a list of three vi,ti tuples
        
        for vi,ti,ni in T:
            key = (vi,ti,ni)
            if key not in vmap:
                #first time we've seen this vertex, so add it
                #to our vertex list and our dictionary
                vdata += [
                    vertexdata[vi][0],
                    vertexdata[vi][1],
                    vertexdata[vi][2],
                    texturedata[ti][0],
                    texturedata[ti][1],
                    normaldata[ni][0],
                    normaldata[ni][1],
                    normaldata[ni][2],
                    vertexdata[vi][3], #w1
                    vertexdata[vi][4], #w2
                    vertexdata[vi][5], #w3
                    vertexdata[vi][6], #w4
                    vertexdata[vi][7], #i1
                    vertexdata[vi][8], #i2
                    vertexdata[vi][9], #i3
                    vertexdata[vi][10], #i4
					0, 0, 0 #tangent placeholder
                ]
                #used for reporting number of vertices.
                #Easier than having to remember to change
                #it every time we modify the vertex format...
                if vsize == 0:
                    vsize = len(vdata)
                    print("Vertex size:",vsize,"floats")

                vmap[key]=(len(vdata)-1)//vsize
            
            idata.append( vmap[key] ) 
    
	#compute tangents
    print("Computing tangents...")
    triIdx = 0
    while triIdx < len(idata):
        temp = []
        for i in range(0, 3):
            temp.append(triIdx)
            triIdx += 1
        
        #q,r,s = triangle vertices
        qrsIdx = [idata[temp[0]] * vsize, idata[temp[1]] * vsize, idata[temp[2]] * vsize]
        q = vdata[qrsIdx[0]:qrsIdx[0]+5]
        r = vdata[qrsIdx[1]:qrsIdx[1]+5]
        s = vdata[qrsIdx[2]:qrsIdx[2]+5]
        qTex = [0, q.pop()]
        rTex = [0, r.pop()]
        sTex = [0, s.pop()]
        
        qTex[0] = q.pop()
        rTex[0] = r.pop()
        sTex[0] = s.pop()
        
        r = sub(r, q) #R'xyz
        s = sub(s, q) #S'xyz
        rTex = sub(rTex, qTex, 2) #R'st
        sTex = sub(sTex, qTex, 2) #S'st
        
        #
        # [Tx Ty Tz]     [R's R't]^-1    [R'x R'y R'z]
        # [Bx By Bz]  =  [S's S't]    *  [S'x S'y S'z]
        # 
        # A = 1/(R's*S't-R't*S's) [S't -R't]
        # 
        # [Tx Ty Tz]  =  A  *  [R'x R'y R'z]
        #                      [S'x S'y S'z]
        #
        
        t = (rTex[0] * sTex[1] - rTex[1] * sTex[0])
        if (t == 0):
            tmp = 0
            print("A tangent was not computed")
        else:
            tmp = 1 / (rTex[0] * sTex[1] - rTex[1] * sTex[0])
        A = [tmp * sTex[1], tmp * -rTex[1]]
        T = [
                 A[0] * r[0] + A[1] * s[0],
                 A[0] * r[1] + A[1] * s[1],
                 A[0] * r[2] + A[1] * s[2]
                 ]
        
        for i in range (0, 3):
            vdata[qrsIdx[0] + (vsize - 3) + i] += T[i]
            vdata[qrsIdx[1] + (vsize - 3) + i] += T[i]
            vdata[qrsIdx[2] + (vsize - 3) + i] += T[i]
        
        ++triIdx
	
    i = 0
    while (i < len(vdata)):
        xyz = vdata[i + (vsize - 3) : i + vsize]
        length = math.sqrt(xyz[0]*xyz[0] + xyz[1]*xyz[1] + xyz[2]*xyz[2])
        for j in range(0, 3):
            if (length > 0):
                vdata[i + (vsize - 3) + j] = xyz[j] / length
            else:
                vdata[i + (vsize - 3) + j] = 0
        i += 19
    
    #information for the user
    print(len(vdata)//vsize,"vertices") 
    print(len(idata)//3,"triangles") 

    #output geometry data
    print("Writing to mesh file")
    ofp = open(outfile,"wb")
    ofp.write(b"mesh_5\n") 
    
    def pad():
        while (ofp.tell()+1) % 4 != 0 :
            ofp.write(b" ")
        ofp.write(b"\n")
        return
        
    #this is the count of the number of floats per vertex * num vertices 
    ofp.write( ( "vertices "+str(len(vdata))+" bytes\n").encode() ) 
    
    #the count of the number of triangle indices
    ofp.write( ( "indices "+str(len(idata))+" bytes\n" ).encode() ) 
    
    #if we know a texture, write it
    if currmtl in mdict and "map_Kd" in mdict[currmtl]:
        ofp.write( ("texture_file "+mdict[currmtl]["map_Kd"]+"\n").encode()  ) 
    if currmtl in mdict and "map_Bump" in mdict[currmtl]:
        ofp.write( ("normal_map "+mdict[currmtl]["map_Bump"]+"\n").encode()  ) 
    if currmtl in mdict and "map_Ks" in mdict[currmtl]:
        ofp.write( ("specular_map "+mdict[currmtl]["map_Ks"]+"\n").encode()  ) 
    
    for obname in objects:
        c=objects[obname]["centroid"]
        nv=objects[obname]["numv"]
        if nv > 0:
            c[0] /= nv
            c[1] /= nv
            c[2] /= nv
        ofp.write(b"centroid "+obname.encode()+b" "+
            (str(c[0])+" "+str(c[1])+" "+str(c[2])).encode()+b"\n")

    #write the vertex data
    ofp.write(b"vertex_data") 
    
    #must pad to 4 byte boundary
    pad()
        
    #convert vertex data to floats and write it
    b = array.array("f",vdata)
    ofp.write(b) 
       
    #write index data
    ofp.write(b"index_data")

    pad() 

    #convert to shorts and write it
    b = array.array("H",idata)
    ofp.write(b) 
    
    
    #now process rig data
    if os.access(inrig,os.F_OK ):
        lst = open(inrig).readlines()
        if lst[0].strip() != "RIG0000":
            print("Bad rig file: starts with",lst[0])
            assert 0
        i=1
        while i < len(lst):
            line = lst[i]
            i+=1
            L = line.strip().split(" ")
            if len(L) == 0:
                pass 
            elif L[0] in ("numbones","numframes","maxdepth","bonenames"):
                ofp.write( (" ".join(L)).encode() )
                ofp.write(b"\n")
            elif L[0] in ("heads","tails","quaternions","translations"):
                numbytes = int(L[1],10)
                mybuffer = []
                ofp.write( (L[0]+" "+str(numbytes)).encode() )
                pad()
                #pull out one quaternion at a time: 16 bytes each
                for j in range(0,numbytes,16):
                    tmp = lst[i].strip()
                    i+=1
                    tmp = tmp.split(" ")
                    for k in range(4):
                        f = float(tmp[k])
                        mybuffer.append(f)
                
                #convert to floats and write
                b = array.array("f",mybuffer)
                ofp.write(b)
            elif L[0] == "matrices":
                numbytes = int(L[1],10)
                mybuffer=[]
                ofp.write( (L[0]+" "+str(numbytes)).encode())
                pad()
                #16 floats per matrix, 4 bytes per float = 48 bytes per matrix
                for j in range(0,numbytes,16*4):
                    tmp = lst[i]
                    i+=1
                    tmp = tmp.split()
                    for k in range(16):
                        f = float(tmp[k])
                        mybuffer.append(f)
                b=array.array("f",mybuffer)
                ofp.write(b)
            else:
                print("Unexpected:",L)
                assert 0

    #indicate we're done
    ofp.write(b"\nend\n") 
    ofp.close()


main()
